File: facebook/src/main/java/com/facebook/share/widget/ShareDialog.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.share.widget;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;

import com.facebook.AccessToken;
import com.facebook.FacebookCallback;
import com.facebook.appevents.AppEventsLogger;
import com.facebook.internal.AnalyticsEvents;
import com.facebook.internal.AppCall;
import com.facebook.internal.CallbackManagerImpl;
import com.facebook.internal.DialogFeature;
import com.facebook.internal.DialogPresenter;
import com.facebook.internal.FacebookDialogBase;
import com.facebook.internal.FragmentWrapper;
import com.facebook.internal.NativeAppCallAttachmentStore;
import com.facebook.internal.Utility;
import com.facebook.share.Sharer;
import com.facebook.share.internal.LegacyNativeDialogParameters;
import com.facebook.share.internal.NativeDialogParameters;
import com.facebook.share.internal.OpenGraphActionDialogFeature;
import com.facebook.share.internal.ShareContentValidation;
import com.facebook.share.internal.ShareDialogFeature;
import com.facebook.share.internal.ShareFeedContent;
import com.facebook.share.internal.ShareInternalUtility;
import com.facebook.share.internal.WebDialogParameters;
import com.facebook.share.model.ShareContent;
import com.facebook.share.model.ShareLinkContent;
import com.facebook.share.model.ShareMediaContent;
import com.facebook.share.model.ShareOpenGraphContent;
import com.facebook.share.model.SharePhoto;
import com.facebook.share.model.SharePhotoContent;
import com.facebook.share.model.ShareVideoContent;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Provides functionality to share content via the Facebook Share Dialog
 */
public final class ShareDialog
        extends FacebookDialogBase<ShareContent, Sharer.Result>
        implements Sharer {

    /**
     * The mode for the share dialog.
     */
    public enum Mode {
        /**
         * The mode is determined automatically.
         */
        AUTOMATIC,
        /**
         * The native dialog is used.
         */
        NATIVE,
        /**
         * The web dialog is used.
         */
        WEB,
        /**
         * The feed dialog is used.
         */
        FEED
    }

    private static final String FEED_DIALOG = "feed";
    public static final String WEB_SHARE_DIALOG = "share";
    private static final String WEB_OG_SHARE_DIALOG = "share_open_graph";

    private static final int DEFAULT_REQUEST_CODE =
            CallbackManagerImpl.RequestCodeOffset.Share.toRequestCode();

    private  boolean shouldFailOnDataError = false;
    // Keep track of Mode overrides for logging purposes.
    private boolean isAutomaticMode = true;

    /**
     * Helper to show the provided {@link com.facebook.share.model.ShareContent} using the provided
     * Activity. No callback will be invoked.
     *
     * @param activity Activity to use to share the provided content
     * @param shareContent Content to share
     */
    public static void show(
            final Activity activity,
            final ShareContent shareContent) {
        new ShareDialog(activity).show(shareContent);
    }

    /**
     * Helper to show the provided {@link com.facebook.share.model.ShareContent} using the provided
     * Fragment. No callback will be invoked.
     *
     * @param fragment android.support.v4.app.Fragment to use to share the provided content
     * @param shareContent Content to share
     */
    public static void show(
            final Fragment fragment,
            final ShareContent shareContent) {
        show(new FragmentWrapper(fragment), shareContent);
    }

    /**
     * Helper to show the provided {@link com.facebook.share.model.ShareContent} using the provided
     * Fragment. No callback will be invoked.
     *
     * @param fragment android.app.Fragment to use to share the provided content
     * @param shareContent Content to share
     */
    public static void show(
            final android.app.Fragment fragment,
            final ShareContent shareContent) {
        show(new FragmentWrapper(fragment), shareContent);
    }

    private static void show(
            final FragmentWrapper fragmentWrapper,
            final ShareContent shareContent) {
        new ShareDialog(fragmentWrapper).show(shareContent);
    }

    /**
     * Indicates whether it is possible to show the dialog for
     * {@link com.facebook.share.model.ShareContent} of the specified type.
     *
     * @param contentType Class of the intended {@link com.facebook.share.model.ShareContent} to
     *                    share.
     * @return True if the specified content type can be shown via the dialog
     */
    public static boolean canShow(Class<? extends ShareContent> contentType) {
        return canShowWebTypeCheck(contentType) || canShowNative(contentType);
    }

    private static boolean canShowNative(Class<? extends ShareContent> contentType) {
        DialogFeature feature = getFeature(contentType);

        return feature != null && DialogPresenter.canPresentNativeDialogWithFeature(feature);
    }

    private static boolean canShowWebTypeCheck(Class<? extends ShareContent> contentType) {
        // If we don't have an instance of a ShareContent, then all we can do is check whether
        // this is a ShareLinkContent, which can be shared if configured properly.
        // The instance method version of this check is more accurate and should be used on
        // ShareDialog instances.

        // SharePhotoContent currently requires the user staging endpoint, so we need a user access
        // token, so we need to see if we have one
        final AccessToken accessToken = AccessToken.getCurrentAccessToken();
        final boolean haveUserAccessToken = accessToken != null && !accessToken.isExpired();

        return ShareLinkContent.class.isAssignableFrom(contentType)
                || ShareOpenGraphContent.class.isAssignableFrom(contentType)
                || (SharePhotoContent.class.isAssignableFrom(contentType) && haveUserAccessToken);
    }

    /**
     * Constructs a new ShareDialog.
     * @param activity Activity to use to share the provided content.
     */
    public ShareDialog(Activity activity) {
        super(activity, DEFAULT_REQUEST_CODE);

        ShareInternalUtility.registerStaticShareCallback(DEFAULT_REQUEST_CODE);
    }

    /**
     * Constructs a new ShareDialog.
     * @param fragment android.support.v4.app.Fragment to use to share the provided content.
     */
    public ShareDialog(Fragment fragment) {
        this(new FragmentWrapper(fragment));

    }

    /**
     * Constructs a new ShareDialog.
     * @param fragment android.app.Fragment to use to share the provided content.
     */
    public ShareDialog(android.app.Fragment fragment) {
        this(new FragmentWrapper(fragment));
    }

    private ShareDialog(FragmentWrapper fragmentWrapper) {
        super(fragmentWrapper, DEFAULT_REQUEST_CODE);

        ShareInternalUtility.registerStaticShareCallback(DEFAULT_REQUEST_CODE);
    }

    // for ShareDialog use only
    ShareDialog(Activity activity, int requestCode) {
        super(activity, requestCode);

        ShareInternalUtility.registerStaticShareCallback(requestCode);
    }

    // for ShareDialog use only
    ShareDialog(Fragment fragment, int requestCode) {
        this(new FragmentWrapper(fragment), requestCode);

    }

    ShareDialog(android.app.Fragment fragment, int requestCode) {
        this(new FragmentWrapper(fragment), requestCode);
    }

    private ShareDialog(FragmentWrapper fragmentWrapper, int requestCode) {
        super(fragmentWrapper, requestCode);

        ShareInternalUtility.registerStaticShareCallback(requestCode);
    }

    @Override
    protected void registerCallbackImpl(
            final CallbackManagerImpl callbackManager,
            final FacebookCallback<Result> callback) {
        ShareInternalUtility.registerSharerCallback(
                getRequestCode(), callbackManager, callback);
    }

    @Override
    public boolean getShouldFailOnDataError() {
        return this.shouldFailOnDataError;
    }

    @Override
    public void setShouldFailOnDataError(boolean shouldFailOnDataError) {
        this.shouldFailOnDataError = shouldFailOnDataError;
    }

    /**
     * Call this to check if the Share Dialog can be shown in a specific mode.
     *
     * @param mode Mode of the Share Dialog
     * @return True if the dialog can be shown in the passed in Mode
     */
    public boolean canShow(ShareContent content, Mode mode) {
        return canShowImpl(content, (mode == Mode.AUTOMATIC) ? BASE_AUTOMATIC_MODE : mode);
    }

    /**
     * Call this to show the Share Dialog in a specific mode
     * @param mode Mode of the Share Dialog
     */
    public void show(ShareContent content, Mode mode) {
        isAutomaticMode = (mode == Mode.AUTOMATIC);

        showImpl(content, isAutomaticMode ? BASE_AUTOMATIC_MODE : mode);
    }

    @Override
    protected AppCall createBaseAppCall() {
        return new AppCall(getRequestCode());
    }

    @Override
    protected List<ModeHandler> getOrderedModeHandlers() {
        ArrayList<ModeHandler> handlers = new ArrayList<>();
        handlers.add(new NativeHandler());
        handlers.add(new FeedHandler()); // Feed takes precedence for link-shares for Mode.AUTOMATIC
        handlers.add(new WebShareHandler());

        return handlers;
    }

    private class NativeHandler extends ModeHandler {
        @Override
        public Object getMode() {
            return Mode.NATIVE;
        }

        @Override
        public boolean canShow(final ShareContent content, boolean isBestEffort) {
            if (content == null) {
                return false;
            }

            boolean canShowResult = true;
            if (!isBestEffort) {
                // The following features are considered best-effort and will not prevent the
                // native share dialog from being presented, even if the installed version does
                // not support the feature.
                // However, to let apps pivot to a different approach or dialog (for example, Web),
                // we need to be able to signal back when native support is lacking.
                if (content.getShareHashtag() != null) {
                    canShowResult = DialogPresenter.canPresentNativeDialogWithFeature(
                            ShareDialogFeature.HASHTAG);
                }
                if ((content instanceof ShareLinkContent) &&
                        (!Utility.isNullOrEmpty(((ShareLinkContent)content).getQuote()))) {
                    canShowResult &= DialogPresenter.canPresentNativeDialogWithFeature(
                            ShareDialogFeature.LINK_SHARE_QUOTES);
                }
            }
            return canShowResult && ShareDialog.canShowNative(content.getClass());
        }

        @Override
        public AppCall createAppCall(final ShareContent content) {
            logDialogShare(getActivityContext(), content, Mode.NATIVE);

            ShareContentValidation.validateForNativeShare(content);

            final AppCall appCall = createBaseAppCall();
            final boolean shouldFailOnDataError = getShouldFailOnDataError();

            DialogPresenter.setupAppCallForNativeDialog(
                    appCall,
                    new DialogPresenter.ParameterProvider() {
                        @Override
                        public Bundle getParameters() {
                            return NativeDialogParameters.create(
                                    appCall.getCallId(),
                                    content,
                                    shouldFailOnDataError);
                        }

                        @Override
                        public Bundle getLegacyParameters() {
                            return LegacyNativeDialogParameters.create(
                                    appCall.getCallId(),
                                    content,
                                    shouldFailOnDataError);
                        }
                    },
                    getFeature(content.getClass()));

            return appCall;
        }
    }

    private class WebShareHandler extends ModeHandler {
        @Override
        public Object getMode() {
            return Mode.WEB;
        }

        @Override
        public boolean canShow(final ShareContent content, boolean isBestEffort) {
            return (content != null) && ShareDialog.canShowWebTypeCheck(content.getClass());
        }

        @Override
        public AppCall createAppCall(final ShareContent content) {
            logDialogShare(getActivityContext(), content, Mode.WEB);

            final AppCall appCall = createBaseAppCall();

            ShareContentValidation.validateForWebShare(content);

            Bundle params;
            if (content instanceof ShareLinkContent) {
                params = WebDialogParameters.create((ShareLinkContent)content);
            } else if (content instanceof SharePhotoContent) {
                final SharePhotoContent photoContent =
                        createAndMapAttachments((SharePhotoContent)content, appCall.getCallId());
                params = WebDialogParameters.create(photoContent);
            } else {
                params = WebDialogParameters.create((ShareOpenGraphContent)content);
            }

            DialogPresenter.setupAppCallForWebDialog(
                    appCall,
                    getActionName(content),
                    params);

            return appCall;
        }

        private String getActionName(ShareContent shareContent) {
            if (shareContent instanceof ShareLinkContent
                    || shareContent instanceof SharePhotoContent) {
                return WEB_SHARE_DIALOG;
            } else if (shareContent instanceof ShareOpenGraphContent) {
                return WEB_OG_SHARE_DIALOG;
            }

            return null;
        }

        private SharePhotoContent createAndMapAttachments(
                final SharePhotoContent content,
                final UUID callId) {
            final SharePhotoContent.Builder contentBuilder =
                    new SharePhotoContent.Builder().readFrom(content);
            final List<SharePhoto> photos = new ArrayList<>();
            final List<NativeAppCallAttachmentStore.Attachment> attachments = new ArrayList<>();
            for (int i = 0; i < content.getPhotos().size(); i++) {
                SharePhoto sharePhoto = content.getPhotos().get(i);
                final Bitmap photoBitmap = sharePhoto.getBitmap();

                if (photoBitmap != null) {
                    NativeAppCallAttachmentStore.Attachment attachment =
                            NativeAppCallAttachmentStore.createAttachment(callId, photoBitmap);
                    sharePhoto = new SharePhoto.Builder()
                            .readFrom(sharePhoto)
                            .setImageUrl(Uri.parse(attachment.getAttachmentUrl()))
                            .setBitmap(null)
                            .build();
                    attachments.add(attachment);
                }

                photos.add(sharePhoto);
            }
            contentBuilder.setPhotos(photos);
            NativeAppCallAttachmentStore.addAttachments(attachments);
            return contentBuilder.build();
        }
    }

    private class FeedHandler extends ModeHandler {
        @Override
        public Object getMode() {
            return Mode.FEED;
        }

        @Override
        public boolean canShow(final ShareContent content, boolean isBestEffort) {
            return (content instanceof ShareLinkContent)
                    || (content instanceof ShareFeedContent);
        }

        @Override
        public AppCall createAppCall(final ShareContent content) {
            logDialogShare(getActivityContext(), content, Mode.FEED);
            AppCall appCall = createBaseAppCall();
            Bundle params;
            if (content instanceof ShareLinkContent) {
                ShareLinkContent linkContent = (ShareLinkContent)content;
                ShareContentValidation.validateForWebShare(linkContent);
                params = WebDialogParameters.createForFeed(linkContent);
            } else {
                ShareFeedContent feedContent = (ShareFeedContent)content;
                params = WebDialogParameters.createForFeed(feedContent);
            }

            DialogPresenter.setupAppCallForWebDialog(
                    appCall,
                    FEED_DIALOG,
                    params);

            return appCall;
        }
    }

    private static DialogFeature getFeature(
            Class<? extends ShareContent> contentType) {
        if (ShareLinkContent.class.isAssignableFrom(contentType)) {
            return ShareDialogFeature.SHARE_DIALOG;
        } else if (SharePhotoContent.class.isAssignableFrom(contentType)) {
            return ShareDialogFeature.PHOTOS;
        } else if (ShareVideoContent.class.isAssignableFrom(contentType)) {
            return ShareDialogFeature.VIDEO;
        } else if (ShareOpenGraphContent.class.isAssignableFrom(contentType)) {
            return OpenGraphActionDialogFeature.OG_ACTION_DIALOG;
        } else if (ShareMediaContent.class.isAssignableFrom(contentType)) {
            return ShareDialogFeature.MULTIMEDIA;
        }
        return null;
    }

    private void logDialogShare(Context context, ShareContent content, Mode mode) {
        String displayType;
        if (isAutomaticMode) {
            mode = Mode.AUTOMATIC;
        }

        switch (mode) {
            case AUTOMATIC:
                displayType = AnalyticsEvents.PARAMETER_SHARE_DIALOG_SHOW_AUTOMATIC;
                break;
            case WEB:
                displayType = AnalyticsEvents.PARAMETER_SHARE_DIALOG_SHOW_WEB;
                break;
            case NATIVE:
                displayType = AnalyticsEvents.PARAMETER_SHARE_DIALOG_SHOW_NATIVE;
                break;
            default:
                displayType = AnalyticsEvents.PARAMETER_SHARE_DIALOG_SHOW_UNKNOWN;
                break;
        }

        String contentType;
        DialogFeature dialogFeature = getFeature(content.getClass());
        if (dialogFeature == ShareDialogFeature.SHARE_DIALOG) {
            contentType = AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_STATUS;
        } else if (dialogFeature == ShareDialogFeature.PHOTOS) {
            contentType = AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_PHOTO;
        } else if (dialogFeature == ShareDialogFeature.VIDEO) {
            contentType = AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_VIDEO;
        } else if (dialogFeature == OpenGraphActionDialogFeature.OG_ACTION_DIALOG) {
            contentType = AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_OPENGRAPH;
        } else {
            contentType = AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_UNKNOWN;
        }

        AppEventsLogger logger = AppEventsLogger.newLogger(context);
        Bundle parameters = new Bundle();
        parameters.putString(
                AnalyticsEvents.PARAMETER_SHARE_DIALOG_SHOW,
                displayType
        );
        parameters.putString(
                AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_TYPE,
                contentType
        );
        logger.logSdkEvent(AnalyticsEvents.EVENT_SHARE_DIALOG_SHOW, null, parameters);
    }
}

File: facebook-common/src/main/java/com/facebook/share/widget/ShareDialog.kt
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.share.widget

import android.app.Activity
import android.content.Context
import android.net.Uri
import android.os.Bundle
import androidx.fragment.app.Fragment
import com.facebook.AccessToken
import com.facebook.FacebookCallback
import com.facebook.FacebookSdk
import com.facebook.appevents.InternalAppEventsLogger
import com.facebook.internal.AnalyticsEvents
import com.facebook.internal.AppCall
import com.facebook.internal.CallbackManagerImpl
import com.facebook.internal.DialogFeature
import com.facebook.internal.DialogPresenter
import com.facebook.internal.FacebookDialogBase
import com.facebook.internal.FragmentWrapper
import com.facebook.internal.NativeAppCallAttachmentStore
import com.facebook.share.Sharer
import com.facebook.share.internal.CameraEffectFeature
import com.facebook.share.internal.LegacyNativeDialogParameters
import com.facebook.share.internal.NativeDialogParameters
import com.facebook.share.internal.ShareContentValidation
import com.facebook.share.internal.ShareDialogFeature
import com.facebook.share.internal.ShareFeedContent
import com.facebook.share.internal.ShareInternalUtility
import com.facebook.share.internal.ShareStoryFeature
import com.facebook.share.internal.WebDialogParameters
import com.facebook.share.model.ShareCameraEffectContent
import com.facebook.share.model.ShareContent
import com.facebook.share.model.ShareLinkContent
import com.facebook.share.model.ShareMediaContent
import com.facebook.share.model.SharePhoto
import com.facebook.share.model.SharePhotoContent
import com.facebook.share.model.ShareStoryContent
import com.facebook.share.model.ShareVideoContent
import java.util.UUID

/** Provides functionality to share content via the Facebook Share Dialog */
open class ShareDialog : FacebookDialogBase<ShareContent<*, *>, Sharer.Result>, Sharer {
  /** The mode for the share dialog. */
  enum class Mode {
    /** The mode is determined automatically. */
    AUTOMATIC,
    /** The native dialog is used. */
    NATIVE,
    /** The web dialog is used. */
    WEB,
    /** The feed dialog is used. */
    FEED
  }

  private var shouldFailOnDataError = false

  // Keep track of Mode overrides for logging purposes.
  private var isAutomaticMode = true

  /**
   * Constructs a new ShareDialog.
   *
   * @param activity Activity to use to share the provided content.
   */
  constructor(activity: Activity) : this(activity, DEFAULT_REQUEST_CODE)

  /**
   * Constructs a new ShareDialog without any context. For androidx to create ActivityResultContract
   * only.
   */
  constructor(requestCode: Int = DEFAULT_REQUEST_CODE) : super(requestCode) {
    ShareInternalUtility.registerStaticShareCallback(requestCode)
  }

  /**
   * Constructs a new ShareDialog.
   *
   * @param fragment androidx.fragment.app.Fragment to use to share the provided content.
   */
  constructor(fragment: Fragment) : this(FragmentWrapper(fragment))

  /**
   * Constructs a new ShareDialog.
   *
   * @param fragment android.app.Fragment to use to share the provided content.
   */
  constructor(fragment: android.app.Fragment) : this(FragmentWrapper(fragment))

  /** for internal use only */
  constructor(activity: Activity, requestCode: Int) : super(activity, requestCode) {
    ShareInternalUtility.registerStaticShareCallback(requestCode)
  }
  /** for internal use only */
  constructor(fragment: Fragment, requestCode: Int) : this(FragmentWrapper(fragment), requestCode)
  /** for internal use only */
  constructor(
      fragment: android.app.Fragment,
      requestCode: Int
  ) : this(FragmentWrapper(fragment), requestCode)
  /** for internal use only */
  constructor(
      fragmentWrapper: FragmentWrapper,
      requestCode: Int = DEFAULT_REQUEST_CODE
  ) : super(fragmentWrapper, requestCode) {
    ShareInternalUtility.registerStaticShareCallback(requestCode)
  }

  override fun registerCallbackImpl(
      callbackManager: CallbackManagerImpl,
      callback: FacebookCallback<Sharer.Result>
  ) {
    ShareInternalUtility.registerSharerCallback(requestCode, callbackManager, callback)
  }

  override fun getShouldFailOnDataError(): Boolean = shouldFailOnDataError

  override fun setShouldFailOnDataError(shouldFailOnDataError: Boolean) {
    this.shouldFailOnDataError = shouldFailOnDataError
  }

  /**
   * Call this to check if the Share Dialog can be shown in a specific mode.
   *
   * @param mode Mode of the Share Dialog
   * @return True if the dialog can be shown in the passed in Mode
   */
  open fun canShow(content: ShareContent<*, *>, mode: Mode): Boolean {
    return canShowImpl(content, if (mode == Mode.AUTOMATIC) BASE_AUTOMATIC_MODE else mode)
  }

  /**
   * Call this to show the Share Dialog in a specific mode
   *
   * @param mode Mode of the Share Dialog
   */
  open fun show(content: ShareContent<*, *>, mode: Mode) {
    isAutomaticMode = mode == Mode.AUTOMATIC
    showImpl(content, if (isAutomaticMode) BASE_AUTOMATIC_MODE else mode)
  }

  override fun createBaseAppCall(): AppCall = AppCall(requestCode)

  // Feed takes precedence for link-shares for Mode.AUTOMATIC
  // Share into story
  override val orderedModeHandlers: List<ModeHandler> =
      arrayListOf(
          NativeHandler(),
          FeedHandler(),
          WebShareHandler(),
          CameraEffectHandler(),
          ShareStoryHandler())

  private inner class NativeHandler : ModeHandler() {
    override var mode: Any = Mode.NATIVE

    override fun canShow(content: ShareContent<*, *>, isBestEffort: Boolean): Boolean {
      if (content is ShareCameraEffectContent || content is ShareStoryContent) {
        return false
      }
      var canShowResult = true
      if (!isBestEffort) {
        // The following features are considered best-effort and will not prevent the
        // native share dialog from being presented, even if the installed version does
        // not support the feature.
        // However, to let apps pivot to a different approach or dialog (for example, Web),
        // we need to be able to signal back when native support is lacking.
        if (content.shareHashtag != null) {
          canShowResult =
              DialogPresenter.canPresentNativeDialogWithFeature(ShareDialogFeature.HASHTAG)
        }
        if (content is ShareLinkContent && !content.quote.isNullOrEmpty()) {
          canShowResult =
              canShowResult &&
                  DialogPresenter.canPresentNativeDialogWithFeature(
                      ShareDialogFeature.LINK_SHARE_QUOTES)
        }
      }
      return canShowResult && canShowNative(content.javaClass)
    }

    override fun createAppCall(content: ShareContent<*, *>): AppCall? {
      logDialogShare(activityContext, content, Mode.NATIVE)
      ShareContentValidation.validateForNativeShare(content)
      val appCall = createBaseAppCall()
      val shouldFailOnDataError = getShouldFailOnDataError()
      val feature = getFeature(content.javaClass) ?: return null
      DialogPresenter.setupAppCallForNativeDialog(
          appCall,
          object : DialogPresenter.ParameterProvider {
            override val parameters: Bundle?
              get() = NativeDialogParameters.create(appCall.callId, content, shouldFailOnDataError)
            override val legacyParameters: Bundle?
              get() =
                  LegacyNativeDialogParameters.create(
                      appCall.callId, content, shouldFailOnDataError)
          },
          feature)
      return appCall
    }
  }

  private inner class WebShareHandler : ModeHandler() {
    override var mode: Any = Mode.WEB

    override fun canShow(content: ShareContent<*, *>, isBestEffort: Boolean): Boolean {
      return canShowWebCheck(content)
    }

    override fun createAppCall(content: ShareContent<*, *>): AppCall? {
      logDialogShare(activityContext, content, Mode.WEB)
      val appCall = createBaseAppCall()
      ShareContentValidation.validateForWebShare(content)
      val params =
          when (content) {
            is ShareLinkContent -> {
              WebDialogParameters.create(content)
            }
            is SharePhotoContent -> {
              val photoContent = createAndMapAttachments(content, appCall.callId)
              WebDialogParameters.create(photoContent)
            }
            else -> {
              return null
            }
          }
      DialogPresenter.setupAppCallForWebDialog(appCall, getActionName(content), params)
      return appCall
    }

    private fun getActionName(shareContent: ShareContent<*, *>): String? {
      return when {
        (shareContent is ShareLinkContent || shareContent is SharePhotoContent) -> WEB_SHARE_DIALOG
        else -> null
      }
    }

    private fun createAndMapAttachments(
        content: SharePhotoContent,
        callId: UUID
    ): SharePhotoContent {
      val contentBuilder = SharePhotoContent.Builder().readFrom(content)
      val photos: MutableList<SharePhoto> = ArrayList()
      val attachments: MutableList<NativeAppCallAttachmentStore.Attachment> = ArrayList()
      for (i in content.photos.indices) {
        var sharePhoto = content.photos[i]
        val photoBitmap = sharePhoto.bitmap
        if (photoBitmap != null) {
          val attachment = NativeAppCallAttachmentStore.createAttachment(callId, photoBitmap)
          sharePhoto =
              SharePhoto.Builder()
                  .readFrom(sharePhoto)
                  .setImageUrl(Uri.parse(attachment.attachmentUrl))
                  .setBitmap(null)
                  .build()
          attachments.add(attachment)
        }
        photos.add(sharePhoto)
      }
      contentBuilder.setPhotos(photos)
      NativeAppCallAttachmentStore.addAttachments(attachments)
      return contentBuilder.build()
    }
  }

  private inner class FeedHandler : ModeHandler() {
    override var mode: Any = Mode.FEED

    override fun canShow(content: ShareContent<*, *>, isBestEffort: Boolean): Boolean {
      return content is ShareLinkContent || content is ShareFeedContent
    }

    override fun createAppCall(content: ShareContent<*, *>): AppCall? {
      logDialogShare(activityContext, content, Mode.FEED)
      val appCall = createBaseAppCall()
      val params =
          when (content) {
            is ShareLinkContent -> {
              ShareContentValidation.validateForWebShare(content)
              WebDialogParameters.createForFeed(content)
            }
            is ShareFeedContent -> {
              WebDialogParameters.createForFeed(content)
            }
            else -> {
              return null
            }
          }
      DialogPresenter.setupAppCallForWebDialog(appCall, FEED_DIALOG, params)
      return appCall
    }
  }

  private inner class CameraEffectHandler : ModeHandler() {
    override var mode: Any = Mode.NATIVE

    override fun canShow(content: ShareContent<*, *>, isBestEffort: Boolean): Boolean {
      val canShowResult = content is ShareCameraEffectContent
      return canShowResult && canShowNative(content.javaClass)
    }

    override fun createAppCall(content: ShareContent<*, *>): AppCall? {
      ShareContentValidation.validateForNativeShare(content)
      val appCall = createBaseAppCall()
      val shouldFailOnDataError = getShouldFailOnDataError()
      val feature = getFeature(content.javaClass) ?: return null
      DialogPresenter.setupAppCallForNativeDialog(
          appCall,
          object : DialogPresenter.ParameterProvider {
            override val parameters: Bundle?
              get() = NativeDialogParameters.create(appCall.callId, content, shouldFailOnDataError)
            override val legacyParameters: Bundle?
              get() =
                  LegacyNativeDialogParameters.create(
                      appCall.callId, content, shouldFailOnDataError)
          },
          feature)
      return appCall
    }
  }

  private inner class ShareStoryHandler : ModeHandler() {
    override var mode: Any = Mode.NATIVE

    override fun canShow(content: ShareContent<*, *>, isBestEffort: Boolean): Boolean {
      val canShowResult = content is ShareStoryContent
      return canShowResult && canShowNative(content.javaClass)
    }

    override fun createAppCall(content: ShareContent<*, *>): AppCall? {
      ShareContentValidation.validateForStoryShare(content)
      val appCall = createBaseAppCall()
      val shouldFailOnDataError = getShouldFailOnDataError()
      val feature = getFeature(content.javaClass) ?: return null
      DialogPresenter.setupAppCallForNativeDialog(
          appCall,
          object : DialogPresenter.ParameterProvider {
            override val parameters: Bundle?
              get() = NativeDialogParameters.create(appCall.callId, content, shouldFailOnDataError)
            override val legacyParameters: Bundle?
              get() =
                  LegacyNativeDialogParameters.create(
                      appCall.callId, content, shouldFailOnDataError)
          },
          feature)
      return appCall
    }
  }

  private fun logDialogShare(context: Context?, content: ShareContent<*, *>, mode: Mode) {
    val displayMode = if (isAutomaticMode) Mode.AUTOMATIC else mode
    val displayType =
        when (displayMode) {
          Mode.AUTOMATIC -> AnalyticsEvents.PARAMETER_SHARE_DIALOG_SHOW_AUTOMATIC
          Mode.WEB -> AnalyticsEvents.PARAMETER_SHARE_DIALOG_SHOW_WEB
          Mode.NATIVE -> AnalyticsEvents.PARAMETER_SHARE_DIALOG_SHOW_NATIVE
          else -> AnalyticsEvents.PARAMETER_SHARE_DIALOG_SHOW_UNKNOWN
        }
    val contentType: String
    val dialogFeature = getFeature(content.javaClass)
    contentType =
        when {
          dialogFeature === ShareDialogFeature.SHARE_DIALOG -> {
            AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_STATUS
          }
          dialogFeature === ShareDialogFeature.PHOTOS -> {
            AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_PHOTO
          }
          dialogFeature === ShareDialogFeature.VIDEO -> {
            AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_VIDEO
          }
          else -> {
            AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_UNKNOWN
          }
        }
    val logger = InternalAppEventsLogger.createInstance(context, FacebookSdk.getApplicationId())
    val parameters = Bundle()
    parameters.putString(AnalyticsEvents.PARAMETER_SHARE_DIALOG_SHOW, displayType)
    parameters.putString(AnalyticsEvents.PARAMETER_SHARE_DIALOG_CONTENT_TYPE, contentType)
    logger.logEventImplicitly(AnalyticsEvents.EVENT_SHARE_DIALOG_SHOW, parameters)
  }

  companion object {
    private val TAG = ShareDialog::class.java.simpleName
    private const val FEED_DIALOG = "feed"
    const val WEB_SHARE_DIALOG = "share"
    private const val WEB_OG_SHARE_DIALOG = "share_open_graph"
    private val DEFAULT_REQUEST_CODE = CallbackManagerImpl.RequestCodeOffset.Share.toRequestCode()

    /**
     * Helper to show the provided [com.facebook.share.model.ShareContent] using the provided
     * Activity. No callback will be invoked.
     *
     * @param activity Activity to use to share the provided content
     * @param shareContent Content to share
     */
    @JvmStatic
    open fun show(activity: Activity, shareContent: ShareContent<*, *>) {
      ShareDialog(activity).show(shareContent)
    }

    /**
     * Helper to show the provided [com.facebook.share.model.ShareContent] using the provided
     * Fragment. No callback will be invoked.
     *
     * @param fragment androidx.fragment.app.Fragment to use to share the provided content
     * @param shareContent Content to share
     */
    @JvmStatic
    open fun show(fragment: Fragment, shareContent: ShareContent<*, *>) {
      show(FragmentWrapper(fragment), shareContent)
    }

    /**
     * Helper to show the provided [com.facebook.share.model.ShareContent] using the provided
     * Fragment. No callback will be invoked.
     *
     * @param fragment android.app.Fragment to use to share the provided content
     * @param shareContent Content to share
     */
    @JvmStatic
    open fun show(fragment: android.app.Fragment, shareContent: ShareContent<*, *>) {
      show(FragmentWrapper(fragment), shareContent)
    }

    private fun show(fragmentWrapper: FragmentWrapper, shareContent: ShareContent<*, *>) {
      ShareDialog(fragmentWrapper).show(shareContent)
    }

    /**
     * Indicates whether it is possible to show the dialog for [ ] of the specified type.
     *
     * @param contentType Class of the intended [com.facebook.share.model.ShareContent] to share.
     * @return True if the specified content type can be shown via the dialog
     */
    @JvmStatic
    open fun canShow(contentType: Class<out ShareContent<*, *>>): Boolean {
      return canShowWebTypeCheck(contentType) || canShowNative(contentType)
    }

    private fun canShowNative(contentType: Class<out ShareContent<*, *>>): Boolean {
      val feature = getFeature(contentType)
      return feature != null && DialogPresenter.canPresentNativeDialogWithFeature(feature)
    }

    private fun canShowWebTypeCheck(contentType: Class<out ShareContent<*, *>>): Boolean {
      // If we don't have an instance of a ShareContent, then all we can do is check whether
      // this is a ShareLinkContent, which can be shared if configured properly.
      // The instance method version of this check is more accurate and should be used on
      // ShareDialog instances.

      // SharePhotoContent currently requires the user staging endpoint, so we need a user access
      // token, so we need to see if we have one
      return (ShareLinkContent::class.java.isAssignableFrom(contentType) ||
          (SharePhotoContent::class.java.isAssignableFrom(contentType) &&
              AccessToken.isCurrentAccessTokenActive()))
    }

    private fun canShowWebCheck(content: ShareContent<*, *>): Boolean {
      if (!canShowWebTypeCheck(content.javaClass)) {
        return false
      }
      return true
    }

    private fun getFeature(contentType: Class<out ShareContent<*, *>>): DialogFeature? {
      return when {
        ShareLinkContent::class.java.isAssignableFrom(contentType) -> {
          ShareDialogFeature.SHARE_DIALOG
        }
        SharePhotoContent::class.java.isAssignableFrom(contentType) -> {
          ShareDialogFeature.PHOTOS
        }
        ShareVideoContent::class.java.isAssignableFrom(contentType) -> {
          ShareDialogFeature.VIDEO
        }
        ShareMediaContent::class.java.isAssignableFrom(contentType) -> {
          ShareDialogFeature.MULTIMEDIA
        }
        ShareCameraEffectContent::class.java.isAssignableFrom(contentType) -> {
          CameraEffectFeature.SHARE_CAMERA_EFFECT
        }
        ShareStoryContent::class.java.isAssignableFrom(contentType) -> {
          ShareStoryFeature.SHARE_STORY_ASSET
        }
        else -> null
      }
    }
  }
}
